### DISCUSSION TITLE
More Type Hints in Code and Docs

### DISCUSSION THREAD

**Discussion by Menelion** [laravel/framework#36595] [Ideas] [answered]
Hi everyone,  
As Laravel is now powered by PHP 7.3 or above, it's safe to add more type hints both in the code and in the docs. This code style seems to be encouraged by core PHP developers as PHP is moving towards stricter typization.

### Example

```php
<?php
// ----- Now -----

    /**
     * Get the login username to be used by the controller.
     *
     * @return string
     */
    public function username()
    {
        return 'email';
    }

// ----- In Future -----

    /**
     * Get the login username to be used by the controller.
     */
    public function username(): string
    {
        return 'email';
    }
```

And, of course, I'm willing to help in this work.

**Comment by allanlaal** [comment]
I agree, any modern IDE will work a lot better with statically typed code
also static analysis (i.e. https://github.com/nunomaduro/larastan/ ) is now a very helpful thing (even with laravel/framework not being completely typed)

sadly some of the top commiters to laravel/framework have not yet started to write typed code - I guess they are used to docblocks

**Comment by allanlaal** [comment]
@nunomaduro what are your thoughts on using (only) static types in Laravel/framework? I noticed a random recent commit of yours still used docblocks instead..

**Comment by shaedrich** [comment, marked as answer]
This was added to Laravel 10 via #44545

**Reply by allanlaal** [reply]
I wouldn't consider Laravel 10 typed, since only a few types were added to core. its not a complete rewrite at all

**Reply by shaedrich** [reply]
It _is_ an improvement, but I agree, it could be better. However, that's probably complicated in some places if one wants to maintain backwards compatibility. Yet, we might get that in the future.

**Reply by allanlaal** [reply]
10 would've been a great Major version nr to start with complete 100% strict typing support. I hope we dont have to wait until 20 then :dagger: 

Major versions are for breaking changes. The stuff in 10 could've been just another 9.x imho

**Reply by shaedrich** [reply]
Sure, if you go by semantic versioning. But Taylor Otwell does something more along the lines of "never change a running system—only make exceptions for people you personally like"
